### PR TITLE
Add four letter commands to keeper

### LINKED ISSUE #28649: Add 4 letters words to NuRaftKeeper
**Use case**

To monitor NuRaftKeeper internal status such as server info, client info, watch info, ephemeral nodes info etc. Zookeeper provides 4 letters admin tool. You can check it [here](https://zookeeper.apache.org/doc/r3.4.8/zookeeperAdmin.html#sc_zkCommands).

**Describe the solution you'd like**

1. When startup register 4 letter word command with a <code, command_executor> map. The code is name reinterpreted int32.
2. When new connection comes, first extract the first 4 characters and check whether it is 4lw command, if yes, run the command.
3. add some methods to `KeeperStorageDispatcher` to   collect server info.

Because the first request of a connection is hand shake request whose first 4 characters is request length and the length is fixed 44 or 45.  All  4lw code is larger than 2^24 or lower than 0, so collision never happens. 

In JD, we already added some 4lw commands to `SvsKeeper` which is a NuRaft based system like`NuRaftKeeper` . If the idea is ok, I'd like to add it to `NuRaftKeeper`

### PR DESCRIPTION
**Pull request by JackyWoo**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

We need to implement similar commands in clickhouse-keeper: https://zookeeper.apache.org/doc/r3.4.8/zookeeperAdmin.html#sc_zkCommands

Detailed description / Documentation draft:

1. When startup register 4 letter word command with a <code, command_executor> map. The code is name reinterpreted int32.
2. When new connection comes, first extract the first 4 characters and check whether it is 4lw command, if yes, run the command.
3. add some methods to KeeperStorageDispatcher to collect server info.

Because the first request of a connection is hand shake request whose first 4 characters is request length and the length is fixed 44 or 45. All 4lw code is larger than 2^24 or lower than 0, so collision never happens.

The pr is still WIP the first commit only add  'ruok' command.

The pr will close #28649 #26976.

### CODE REVIEW

File: tests/jepsen.clickhouse-keeper/docker/control/jepsen/target/classes/META-INF/maven/jepsen.keeper/jepsen.keeper/pom.properties [hunk 1]
```diff
@@ -0,0 +1,3 @@
+groupId=jepsen.keeper
```
Comment (alesapin) [inline tests/jepsen.clickhouse-keeper/docker/control/jepsen/target/classes/META-INF/maven/jepsen.keeper/jepsen.keeper/pom.properties:1]: accident changes?

Comment (JackyWoo) [reply tests/jepsen.clickhouse-keeper/docker/control/jepsen/target/classes/META-INF/maven/jepsen.keeper/jepsen.keeper/pom.properties:1]: It is an accident changes.

**Comment by JackyWoo** [comment]
@alesapin if the pr is ok, I will work on the rest part.

File: src/Coordination/FourLetterCommand.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+#include <Interpreters/Context_fwd.h>
+#include <IO/WriteBufferFromString.h>
+#include <common/types.h>
+
+namespace DB
+{
+struct FourLetterCommand;
+using FourLetterCommandPtr = std::shared_ptr<DB::FourLetterCommand>;
+
+struct FourLetterCommand
```
Comment (alesapin) [inline src/Coordination/FourLetterCommand.h:16]: It's an interface, so let's name it `IFourLetterCommand`

File: src/Coordination/FourLetterCommand.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+struct RuokCommand : public FourLetterCommand
+{
+    explicit RuokCommand(const Context & global_context_) : FourLetterCommand(global_context_) { }
+
+    String name() override;
+    void run(String & res) override;
+    ~RuokCommand() override;
+};
+
+struct FourLetterCommands
```
Comment (alesapin) [inline src/Coordination/FourLetterCommand.h:51]: It looks like factory. We have a lot of them in our code. Take a look at StorageFactory + registerStorages.

File: src/Coordination/FourLetterCommand.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+namespace DB
+{
+struct FourLetterCommand;
+using FourLetterCommandPtr = std::shared_ptr<DB::FourLetterCommand>;
+
+struct FourLetterCommand
+{
+public:
+    using StringBuffer = DB::WriteBufferFromOwnString;
+    explicit FourLetterCommand(const Context & global_context_);
```
Comment (alesapin) [inline src/Coordination/FourLetterCommand.h:20]: Why we need global_context in all commands? Maybe KeeperDispatcher?

File: src/Coordination/FourLetterCommand.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    virtual String name() = 0;
+    virtual void run(String & res) = 0;
```
Comment (alesapin) [inline src/Coordination/FourLetterCommand.h:23]: Maybe it would be simpler to return String then "return parameter"?

**Review by alesapin — COMMENTED** [review]
In general LGTM. Let's implement several more interesting commands!

**Comment by alesapin** [comment]
@JackyWoo How is it going?)

**Comment by JackyWoo** [comment]
@alesapin Sorry for late reply. Because We are in the biggest event of the year, the work delayed. I will add more commands in two or three weaks.

**Review by alesapin — COMMENTED** [review]
Cool. Are you going to add more commands in this PR? Or we can merge it and continue in a new one?

**Comment by JackyWoo** [comment]
@alesapin add `sizeOf` util functions to SnapshotableHashTable which were chosen at compile time according to where element has a size method.

**Comment by JackyWoo** [comment]
add all zookeeper 4lw commands except gtmk and stmk

File: src/Coordination/KeeperStorage.cpp [hunk 6]
```diff
@@ -1220,4 +1221,62 @@ void KeeperStorage::clearDeadWatches(int64_t session_id)
...
+        for (const String & obj : objs)
+        {
+            buf << "\t" << obj << "\n";
+        }
+    };
+
+    for (const auto & e : sessions_and_watchers)
+    {
+        buf << "0x" << getHexUIntLowercase(e.first) << "\n";
+        write_str_set(e.second);
```
Comment (alesapin) [inline src/Coordination/KeeperStorage.cpp:1237]: Only one usage. Does it make sense to create lambda?

File: src/Coordination/KeeperStorage.h [hunk 7]
```diff
@@ -39,6 +39,17 @@ class KeeperStorage
         Coordination::Stat stat{};
         int32_t seq_num = 0;
         ChildrenSet children{};
+
+        /// object memory size
+        UInt64 size() const
```
Comment (alesapin) [inline src/Coordination/KeeperStorage.h:44]: sizeInBytes?

File: src/Coordination/SnapshotableHashTable.h [hunk 8]
```diff
@@ -15,6 +15,19 @@ struct ListNode
     bool active_in_map;
 };
 
+template <class V>
```
Comment (alesapin) [inline src/Coordination/SnapshotableHashTable.h:18]: Too complex to be honest. Maybe just omit this check? Or add static assert in the constructor of SnapshotableHashTable?

File: src/Coordination/SnapshotableHashTable.h [hunk 9]
```diff
@@ -28,6 +41,96 @@ class SnapshotableHashTable
...
+                {
+                    approximate_data_size += key_size;
+                    approximate_data_size += value_size;
+                    if (!snapshot_mode)
+                    {
+                        approximate_data_size += key_size;
+                        approximate_data_size -= old_value_size;
+                    }
+                }
+                /// inseert
```
Comment (alesapin) [inline src/Coordination/SnapshotableHashTable.h:85]: typo

File: src/Server/KeeperTCPHandler.cpp [hunk 10]
```diff
@@ -418,6 +494,169 @@ std::pair<Coordination::OpNum, Coordination::XID> KeeperTCPHandler::receiveReque
...
+                response->zxid,
+                Poco::Timestamp().epochMicroseconds() / 1000,
+                elapsed);
+        }
+    }
+}
+
+void KeeperTCPHandler::dumpStats(WriteBufferFromOwnString & buf, bool brief)
+{
+    auto write_str = [&buf](const String & str) { buf.write(str.data(), str.size()); };
```
Comment (alesapin) [inline src/Server/KeeperTCPHandler.cpp:532]: why not just helper from `IO/WriteHelpers.h`?

File: src/Coordination/KeeperStateManager.cpp [hunk 11]
```diff
@@ -11,27 +11,6 @@ namespace ErrorCodes
     extern const int RAFT_ERROR;
 }
 
-namespace
```
Comment (alesapin) [inline src/Coordination/KeeperStateManager.cpp:14]: Why this function disappeared?

File: src/Coordination/KeeperDispatcher.cpp [hunk 12]
```diff
@@ -14,9 +18,36 @@ namespace ErrorCodes
     extern const int SYSTEM_ERROR;
 }
 
+using Poco::Path;
```
Comment (alesapin) [inline src/Coordination/KeeperDispatcher.cpp:21]: We prefer to use `std::filesystem` Poco::Path banned in our codebase...

File: src/Coordination/FourLetterCommand.cpp [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+    write("Zxid", to_string(state_machine.getLastProcessedZxid()));
+    write("Mode", keeper_info.getRole());
+    write("Node count", to_string(state_machine.getNodeCount()));
+
+    return buf.str();
+}
+
+String StatCommand::run()
+{
+    using std::to_string;
```
Comment (alesapin) [inline src/Coordination/FourLetterCommand.cpp:329]: better `toString`

File: src/Coordination/CoordinationSettings.h [hunk 14]
```diff
@@ -51,4 +52,35 @@ struct CoordinationSettings : public BaseSettings<CoordinationSettingsTraits>
 
 using CoordinationSettingsPtr = std::shared_ptr<CoordinationSettings>;
 
+struct KeeperSettings
```
Comment (alesapin) [inline src/Coordination/CoordinationSettings.h:55]: Better to add comment.

**Review by alesapin — COMMENTED** [review]
Incredible job is done, thank you! Almost all my remarks are about style.

**Comment by JackyWoo** [comment]
Hi alesapin, really appreciate for your help.

File: src/Coordination/SnapshotableHashTable.h (hunk 8 above)
Comment (JackyWoo) [reply src/Coordination/SnapshotableHashTable.h:18]: The main purpose of the check is to make the unit test compatible. If you still think it is to complex, I will remove it.

File: src/Coordination/KeeperStateManager.cpp (hunk 11 above)
Comment (JackyWoo) [reply src/Coordination/KeeperStateManager.cpp:14]: move it to `CoordinationSettings.h`

File: src/Coordination/KeeperDispatcher.cpp (hunk 12 above)
Comment (JackyWoo) [reply src/Coordination/KeeperDispatcher.cpp:21]: May I ask why Poco::Path is banned, for code consistency or something else?

Comment (alesapin) [reply src/Coordination/KeeperDispatcher.cpp]: For code consistency and less dependencies.

**Review by alesapin — COMMENTED** [review]
@JackyWoo Thanks! I'll slightly refactor code and merge it. If you are going to add something else -- fill free, I'll try to resolve conflicts.

**Comment by JackyWoo** [comment]
Ok, thank you.  The work is almost completed, but there are some CI checking errors, I'am trying to fix it.

**Comment by JackyWoo** [comment]
hi alesapin, I am new to ClickHouse and C/C++. May I ask which Ubuntu version is recommended for developing and if I want to deploy CI in gitlab whether there is a ready-made docs. 

Thank you in advance.

**Comment by alesapin** [comment]
> hi alesapin, I am new to ClickHouse and C/C++. May I ask which Ubuntu version is recommended for developing and if I want to deploy CI in gitlab whether there is a ready-made docs.
> 
> Thank you in advance.

I recommend to start with https://clickhouse.com/docs/en/development/developer-instruction/

**Comment by alesapin** [comment]
I'll also try to fix checks.

**Comment by JackyWoo** [comment]
Ok, thank you.

File: src/Common/getCurrentProcessFDCount.cpp [hunk 15]
```diff
@@ -1,35 +1,35 @@
...
+    auto proc_fs_path = fmt::format("/proc/{}/fd", pid);
+    if (fs::exists(proc_fs_path))
+    {
+        result = std::distance(fs::directory_iterator(proc_fs_path), fs::directory_iterator{});
+    }
+    else if (fs::exists("/dev/fd"))
     {
-        readIntText(result, command->out);
-        command->wait();
+        result = std::distance(fs::directory_iterator("/dev/fd"), fs::directory_iterator{});
```
Comment (JackyWoo) [inline src/Common/getCurrentProcessFDCount.cpp:26]: Very nice work, thanks.

**Review by alesapin — APPROVED** [review]

File: src/Coordination/FourLetterCommand.cpp [hunk 16]
```diff
@@ -143,18 +143,17 @@ bool FourLetterCommandFactory::isEnabled(int32_t code)
...
-    using Poco::StringTokenizer;
     const auto & keeper_settings = keeper_dispatcher.getKeeperConfigurationAndSettings();
 
     String list_str = keeper_settings->four_letter_word_white_list;
-    StringTokenizer tokenizer(list_str, ",", 2);
+    Strings tokens;
+    splitInto<','>(tokens, list_str);
 
-    for (const String & token : tokenizer)
+    for (const String & token : tokens)
```
Comment (JackyWoo) [inline src/Coordination/FourLetterCommand.cpp:151]: trim token first

File: tests/queries/shell_config.sh [hunk 17]
```diff
@@ -73,6 +73,8 @@ export CLICKHOUSE_PORT_MYSQL=${CLICKHOUSE_PORT_MYSQL:=$(${CLICKHOUSE_EXTRACT_CON
 export CLICKHOUSE_PORT_MYSQL=${CLICKHOUSE_PORT_MYSQL:="9004"}
 export CLICKHOUSE_PORT_POSTGRESQL=${CLICKHOUSE_PORT_POSTGRESQL:=$(${CLICKHOUSE_EXTRACT_CONFIG} --try --key=postgresql_port 2>/dev/null)} 2>/dev/null
 export CLICKHOUSE_PORT_POSTGRESQL=${CLICKHOUSE_PORT_POSTGRESQL:="9005"}
+export CLICKHOUSE_PORT_POSTGRESQL=${CLICKHOUSE_PORT_POSTGRESQL:=$(${CLICKHOUSE_EXTRACT_CONFIG} --try --key=keeper_server.tcp_port 2>/dev/null)} 2>/dev/null
```
Comment (JackyWoo) [inline tests/queries/shell_config.sh:76]: CLICKHOUSE_PORT_KEEPER?

**Review by JackyWoo — COMMENTED** [review]
Very nice work, thanks.

File: tests/queries/shell_config.sh (hunk 17 above)
Comment (alesapin) [reply tests/queries/shell_config.sh:76]: Misstype :)

File: src/Coordination/FourLetterCommand.cpp (hunk 16 above)
Comment (alesapin) [reply src/Coordination/FourLetterCommand.cpp]: Done

**Review by JackyWoo — COMMENTED** [review]
approved

File: src/Coordination/KeeperDispatcher.cpp [hunk 18]
```diff
@@ -577,4 +583,67 @@ void KeeperDispatcher::updateConfiguration(const Poco::Util::AbstractConfigurati
...
+
+uint64_t KeeperDispatcher::getSnapDirSize() const
+{
+    return getDirSize(configuration_and_settings->snapshot_storage_path);
+}
+
+Keeper4LWInfo KeeperDispatcher::getKeeper4LWInfo() const
+{
+    Keeper4LWInfo result;
+    result.is_standalone = !result.is_follower && server->getFollowerCount() == 0;
```
[the remark this answers is not in the document]
Comment (alesapin) [reply src/Coordination/KeeperDispatcher.cpp:626]: But theoretically we can be standalone follower...

File: src/Coordination/tests/gtest_coordination.cpp [hunk 19]
```diff
@@ -983,24 +983,24 @@ TEST_P(CoordinationTest, SnapshotableHashMapDataSize)
     /// 160 = sizeof Node
```
Comment (JackyWoo) [inline src/Coordination/tests/gtest_coordination.cpp:983]: Fix or delete comment

**Comment by alesapin** [comment]
@mergifyio update

**Comment by alesapin** [comment]
Finally fixed:
```
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.659512 [ 26617 ] {} <Fatal> BaseDaemon: ########################################
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.666285 [ 26617 ] {} <Fatal> BaseDaemon: (version 21.12.1.8825, build id: 7EAA75FE01C502B751BD36D980EEF0442297DA43) (from thread 2697) (no query) Received signal Aborted (6)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.670013 [ 26617 ] {} <Fatal> BaseDaemon: 
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.672388 [ 26617 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f2f5414218b 0x7f2f54121859 0x1c06502e 0x209e3b08 0x209e3a75 0xd3aa51b 0x1cc42f59 0xd3bd04d 0xd3bcf85 0x1debd12b 0x1debd2ee 0x1ed33e2e 0x1ee27c27 0x1ee254ec 0x7f2f542f7609 0x7f2f5421e293
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.673487 [ 26617 ] {} <Fatal> BaseDaemon: 3. raise @ 0x4618b in /usr/lib/x86_64-linux-gnu/libc-2.31.so
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.683707 [ 26617 ] {} <Fatal> BaseDaemon: 4. abort @ 0x25859 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.701779 [ 26617 ] {} <Fatal> BaseDaemon: 5. ./obj-x86_64-linux-gnu/../base/daemon/BaseDaemon.cpp:454: terminate_handler() @ 0x1c06502e in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.721355 [ 26617 ] {} <Fatal> BaseDaemon: 6. ./obj-x86_64-linux-gnu/../contrib/libcxxabi/src/cxa_handlers.cpp:61: std::__terminate(void (*)()) @ 0x209e3b08 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:49.729291 [ 26617 ] {} <Fatal> BaseDaemon: 7. ./obj-x86_64-linux-gnu/../contrib/libcxxabi/src/cxa_handlers.cpp:0: std::terminate() @ 0x209e3a75 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.229384 [ 26617 ] {} <Fatal> BaseDaemon: 8. ? @ 0xd3aa51b in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.278621 [ 26617 ] {} <Fatal> BaseDaemon: 9. ./obj-x86_64-linux-gnu/../src/IO/WriteBufferFromPocoSocket.cpp:89: ? @ 0x1cc42f59 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.361595 [ 26617 ] {} <Fatal> BaseDaemon: 10. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2474: std::__1::__shared_count::__release_shared() @ 0xd3bd04d in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.393948 [ 26617 ] {} <Fatal> BaseDaemon: 11. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2517: std::__1::__shared_weak_count::__release_shared() @ 0xd3bcf85 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.492828 [ 26617 ] {} <Fatal> BaseDaemon: 12. ./obj-x86_64-linux-gnu/../src/Server/KeeperTCPHandler.cpp:632: DB::KeeperTCPHandler::~KeeperTCPHandler() @ 0x1debd12b in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.630555 [ 26617 ] {} <Fatal> BaseDaemon: 13. ./obj-x86_64-linux-gnu/../src/Server/KeeperTCPHandler.cpp:630: DB::KeeperTCPHandler::~KeeperTCPHandler() @ 0x1debd2ee in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.665356 [ 26617 ] {} <Fatal> BaseDaemon: 14.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/atomic:1072: int std::__1::__cxx_atomic_fetch_sub<int>(std::__1::__cxx_atomic_base_impl<int>*, int, std::__1::memory_order)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.678260 [ 26617 ] {} <Fatal> BaseDaemon: 14.2. inlined from ../contrib/libcxx/include/atomic:1725: std::__1::__atomic_base<int, true>::fetch_sub(int, std::__1::memory_order)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.680215 [ 26617 ] {} <Fatal> BaseDaemon: 14.3. inlined from ../contrib/libcxx/include/atomic:1760: std::__1::__atomic_base<int, true>::operator--()
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.688048 [ 26617 ] {} <Fatal> BaseDaemon: 14.4. inlined from ../contrib/poco/Net/src/TCPServerDispatcher.cpp:236: Poco::Net::TCPServerDispatcher::endConnection()
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.692676 [ 26617 ] {} <Fatal> BaseDaemon: 14. ../contrib/poco/Net/src/TCPServerDispatcher.cpp:119: Poco::Net::TCPServerDispatcher::run() @ 0x1ed33e2e in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.710864 [ 26617 ] {} <Fatal> BaseDaemon: 15. ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:213: Poco::PooledThread::run() @ 0x1ee27c27 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.734268 [ 26617 ] {} <Fatal> BaseDaemon: 16.1. inlined from ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/include/Poco/SharedPtr.h:156: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable> >::assign(Poco::Runnable*)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.739546 [ 26617 ] {} <Fatal> BaseDaemon: 16.2. inlined from ../contrib/poco/Foundation/include/Poco/SharedPtr.h:208: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable> >::operator=(Poco::Runnable*)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.756160 [ 26617 ] {} <Fatal> BaseDaemon: 16. ../contrib/poco/Foundation/src/Thread_POSIX.cpp:360: Poco::ThreadImpl::runnableEntry(void*) @ 0x1ee254ec in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.758285 [ 26617 ] {} <Fatal> BaseDaemon: 17. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.767138 [ 26617 ] {} <Fatal> BaseDaemon: 18. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.948280 [ 26616 ] {} <Fatal> BaseDaemon: 8. ? @ 0xd3aa51b in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:52.964452 [ 26616 ] {} <Fatal> BaseDaemon: 9. ./obj-x86_64-linux-gnu/../src/IO/WriteBufferFromPocoSocket.cpp:89: ? @ 0x1cc42f59 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.020427 [ 26616 ] {} <Fatal> BaseDaemon: 10. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2474: std::__1::__shared_count::__release_shared() @ 0xd3bd04d in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.078542 [ 26616 ] {} <Fatal> BaseDaemon: 11. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2517: std::__1::__shared_weak_count::__release_shared() @ 0xd3bcf85 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.106284 [ 26616 ] {} <Fatal> BaseDaemon: 12. ./obj-x86_64-linux-gnu/../src/Server/KeeperTCPHandler.cpp:632: DB::KeeperTCPHandler::~KeeperTCPHandler() @ 0x1debd12b in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.146149 [ 26616 ] {} <Fatal> BaseDaemon: 13. ./obj-x86_64-linux-gnu/../src/Server/KeeperTCPHandler.cpp:630: DB::KeeperTCPHandler::~KeeperTCPHandler() @ 0x1debd2ee in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.161391 [ 26616 ] {} <Fatal> BaseDaemon: 14.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/atomic:1072: int std::__1::__cxx_atomic_fetch_sub<int>(std::__1::__cxx_atomic_base_impl<int>*, int, std::__1::memory_order)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.163655 [ 26616 ] {} <Fatal> BaseDaemon: 14.2. inlined from ../contrib/libcxx/include/atomic:1725: std::__1::__atomic_base<int, true>::fetch_sub(int, std::__1::memory_order)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.165854 [ 26616 ] {} <Fatal> BaseDaemon: 14.3. inlined from ../contrib/libcxx/include/atomic:1760: std::__1::__atomic_base<int, true>::operator--()
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.170214 [ 26616 ] {} <Fatal> BaseDaemon: 14.4. inlined from ../contrib/poco/Net/src/TCPServerDispatcher.cpp:236: Poco::Net::TCPServerDispatcher::endConnection()
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.172304 [ 26616 ] {} <Fatal> BaseDaemon: 14. ../contrib/poco/Net/src/TCPServerDispatcher.cpp:119: Poco::Net::TCPServerDispatcher::run() @ 0x1ed33e2e in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.200445 [ 26616 ] {} <Fatal> BaseDaemon: 15. ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:213: Poco::PooledThread::run() @ 0x1ee27c27 in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.217797 [ 26616 ] {} <Fatal> BaseDaemon: 16.1. inlined from ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/include/Poco/SharedPtr.h:156: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable> >::assign(Poco::Runnable*)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.218933 [ 26616 ] {} <Fatal> BaseDaemon: 16.2. inlined from ../contrib/poco/Foundation/include/Poco/SharedPtr.h:208: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable> >::operator=(Poco::Runnable*)
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.220526 [ 26616 ] {} <Fatal> BaseDaemon: 16. ../contrib/poco/Foundation/src/Thread_POSIX.cpp:360: Poco::ThreadImpl::runnableEntry(void*) @ 0x1ee254ec in /usr/bin/clickhouse
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.222241 [ 26616 ] {} <Fatal> BaseDaemon: 17. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.225577 [ 26616 ] {} <Fatal> BaseDaemon: 18. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.272362 [ 26617 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: 56E0317DB6CEECE62B1AAA9078259F89. There is no information about the reference checksum.
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:25:53.634259 [ 26616 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: 56E0317DB6CEECE62B1AAA9078259F89. There is no information about the reference checksum.
/var/log/clickhouse-server/clickhouse-server.log.2:2021.11.22 20:26:14.005281 [ 477 ] {} <Fatal> Application: Child process was terminated by signal 6.```

**Comment by alesapin** [comment]
image
Bugs in old infrastructure

**Pull request merged**